design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,156 @@
+
+This problem can be solved by the implementation of "strongly typed" resource classes. Instead of every resource class being of `kind: ResourceClass`, they would map one-to-one with the concrete managed resource kind for which they provided specifications (i.e. `kind: RDSInstanceClass`).
+
+Unfortunately, this makes it much more difficult to identify default resource classes because the default class controller currently watches for objects of `kind: ResourceClass`. With an unknown number of possible kinds of resource classes that could be installed and serve as default for a claim `kind`, a default class controller cannot determine all class kinds to watch.
+
+One solution would instead be to map resource classes one-to-one with claim kinds (i.e. `MySQLInstanceClass`), but that results in the same issue of varying required parameters that is currently being experienced.
+
+## Proposal
+
+In order to satisfy both the requirements of strong typing and default resource classes, a new custom resource kind, `Defaulter`, can be implemented to specify when a specific resource class kind can serve as default for a given claim kind. The workflow would proceed as follows:
+
+1. An admin installs Crossplane, which specifies the portable claim kinds that are available as part of the core Crossplane.
+2. An admin installs an infrastructure stack (i.e. a set of CRDs and controllers for a provider's managed service offerings).
+3. An admin creates an instance of an installed resource class CRD (i.e. `kind: RDSInstanceClass`). It could look as follows:
+
+```yaml
+apiVersion: database.aws.crossplane.io/v1alpha1
+kind: RDSInstanceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+class: db.t2.small
+masterUsername: masteruser
+securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
+size: "20"
+provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1
+providerRef:
+  name: aws-provider
+reclaimPolicy: Delete
+```
+
+4. An admin creates a new `Defaulter` object, specifying that the previously created resource class (`kind: RDSInstanceClass`) should serve as default for a given claim kind `mysqlinstance.storage.crossplane.io`. A `Defaulter` could look as follows:
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: Defaulter
+metadata:
+  name: mysql-aws-defaulter
+  namespace: crossplane-system
+  labels:
+    defaultFor: mysqlinstance.storage.crossplane.io
+classRef:
+  gvk: rdsinstanceclass.database.aws.crossplane.io/v1alpha1
```

Comment:
Yes that was my intention. If I changed `gvk` to `kind` I believe that would satisfy the object reference and give us enough information to find the resource class. Just `name` and `namespace` would not be enough here because unlike the current implementation of claim's `classRef` we don't know that it is of kind `ResourceClass`. Does my logic seem sound here?

Reply:
That seems sound!